main.cpp

```diff
@@ -0,0 +1,95 @@
+#include <iostream>
+#include <vector>
+#include <string>
+#include <sstream>
+#include <unistd.h>
+#include <sys/wait.h>
+#include <cstdlib>
+#include <fstream>
+using namespace std ;
+
+ofstream history_file("history.txt", ios::app);
+
+vector<string> parse_command(const string& input) {
+    istringstream iss(input);
+    vector<string> args;
+    string token;
+    while (iss >> token) {
+        args.push_back(token);
+    }
+    return args;
+}
+
+string expand_env_variables(const string& input) {
+    string result = input;
+    size_t pos = 0;
+    
+    while ((pos = result.find('$', pos)) != string::npos) {
+        size_t end = result.find_first_of(" \t\n", pos+1);
+        string var_name = result.substr(pos+1, (end == string::npos ? result.length() : end) - pos - 1);
+        const char* var_value = getenv(var_name.c_str());
+        if (var_value) {
+            result.replace(pos, var_name.length() + 1, var_value);
+        }
+        pos += var_name.length() + (var_value ? strlen(var_value) : 0);
+    }
+    return result;
+}
+
+
+
+
+void execute_command(const vector<string>& args, bool run_in_background) {
+    vector<char*> c_args(args.size() + 1); 
+    for (size_t i = 0; i < args.size(); ++i) {
+    
+        c_args[i] = const_cast<char*>(args[i].c_str());
+    }
+    c_args[args.size()] = nullptr; 
+
+    pid_t pid = fork();
+    
+    if (pid == 0) { 
+        execvp(c_args[0], c_args.data());
```

Comment:
`execvp` finds the executable file for you, but you were asked to implement that by yourself.